src/WebSocketServer/v1/BlockchainMethods.cs

```diff
@@ -0,0 +1,251 @@
+using Akka.Actor;
+using Neo.IO;
+using Neo.Json;
+using Neo.Ledger;
+using Neo.Network.P2P;
+using Neo.Network.P2P.Payloads;
+using Neo.SmartContract;
+using Neo.SmartContract.Native;
+using Neo.VM;
+using System;
+using System.Linq;
+using System.Net.WebSockets;
+using static Neo.Ledger.Blockchain;
+
+namespace Neo.Plugins.WebSocketServer.v1
+{
+    internal class BlockchainMethods
+    {
+        private readonly NeoSystem _neoSystem;
+        private readonly LocalNode _localNode;
+
+        public BlockchainMethods(
+            NeoSystem neoSystem)
+        {
+            _neoSystem = neoSystem;
+            _localNode = _neoSystem.LocalNode.Ask<LocalNode>(new LocalNode.GetInstance()).Result;
+            WebSocketServerPlugin.RegisterMethods(this);
+        }
+
+        [WebSocketMethod]
+        public JToken GetBlockHeader(JArray _params)
```

Comment:
Go back and read main comment. As you know PRs are slow. The plain is to just add `[WebSocketMethod]` to the `RPC` methods. However, we are not doing that because `RpcServer` is outdated and unmaintained. We are going to be build a central library to handle all the common methods. Instead of fixing `RpcServer` poor logic structure. I did so here, in this PR. Copy have to copy out `api` methods to tht central library. With this implementation of `WebSockets` you remove the need for running your own node. You just connect and that its. It has its own protocol built-in like the `nodes` P2P, but is not P2P. This plugin sends all `node` events to the client. No need for subscriptions. Could in theory use the data for syncing. All this is in real-time. Plus this is `core` is it not? That's like saying just cause `coz` added and had a `RestServer` 1st we should follow their logic. I waited a long time for `roman-khimov ` to reply and i never heard anything back.... I'll leave it as that for now.

Reply:
> RpcServer is outdated and unmaintained.

Really? Where exactly it's not maintained and how come it's outdated?

> With this implementation of WebSockets you remove the need for running your own node

Different problem not related to WS/JSON-RPC/REST/Protobuf/whatever.

> This plugin sends all node events to the client

Which is a problem right away: clients must be able to choose things they're interested in, public servers shouldn't be bothered with sending everything to everyone.

I don't see any value in this implementation, sorry. We do have well-established JSON-RPC interface. By "well-established" I mean that all of the community (wallets/SDKs/whatever from COZ/AxLabs/etc) knows how to interact with it and has a lot of tooling built around it. And then there are dApps built around that tooling. This has a lot of value.

Now for whatever reason you're trying to tell us that JSON-RPC interface is outdated/unmaintained, we should throw it away and replace with something from this PR. Does it mean that all of the tooling built around the JSON-RPC suddenly turns to dust? Should be rewrite everything, including all of the dApp backends? What's the value for users in doing so? What's gonna be the next rewrite? REST/protobuf/anything else?

https://github.com/nspcc-dev/neo-go/blob/master/docs/notifications.md uses a different approach, it's 100% compatible with the current JSON-RPC (JSON-RPC itself is transport-agnostic, so this costs nothing) and gives some new abilities. Extending existing tooling for it is trivial and there are applications already using it (the extension itself dates back to Legacy days). Which doesn't make it a standard, sure, you can create ten other variations around this theme, but please think about the value you're trying to provide to people.

See https://github.com/neo-project/neo/issues/2853 also.

Reply:
I am in agreement with the idea of being 100% compatible with RPC plugin.

It is the same for Rest Plugin as well, the best scenario is that we have a common set of methods used by these interfaces.

Reply:
@vncoelho 
Ee are merging this repo to `neo` monorepo. Me and @Jim8y where going to do one library for all common logic. But after the `RPC` proposal changes _PR_. That's why I programed the logic kind of the same with `RpcServer` `RpcMethodName` attribute. So you can just add `WebSocketMethod` and be done. Also we can update all the other `Plugins` too; just by adding `WebSocketMethod` attribute.

Reply:
>I don't see any value in this implementation, sorry. We do have well-established JSON-RPC interface. By "well-established" I mean that all of the community (wallets/SDKs/whatever from COZ/AxLabs/etc) knows how to interact with it and has a lot of tooling built around it. And then there are dApps built around that tooling. This has a lot of value.

Limiting the community and developers isn't going to make anything better. We need a good solid reason why with `pros` and `cons`. Other than your feelings why something shouldn't or should be. This PR isn't about `RPC-JSON`; it's about `Websockets`.

>Which is a problem right away: clients must be able to choose things they're interested in, public servers shouldn't be bothered with sending everything to everyone.

Lets overload the server with `HTTP` `POST` requests or with logic that creates events and filters data, so someone can `DDOS` Or have millions of `subscription` events. That would be `x` times each client per `x` number of clients vs the `x` servers. Do the math; it's simpler and cheaper with bandwidth then `CPU` time or `RAM`. 

>Now for whatever reason you're trying to tell us that JSON-RPC interface is outdated/unmaintained, we should throw it away and replace with something from this PR

No one said that. You miss read it or misunderstand. So let me clear that up. We still need `RPC-JSON`. The code in this repo is outdated. Look how long it takes to get a `RPC-JSON` bug merged. Some of those clients or users you talk about using `RPC-JSON` had to wait months for a bug fix to merge for #812; that should of had a `hotfix` within two weeks or reasonable timeframe.

This PR isn't meant to replace `RPC-JSON` or `RestServer`. It meant for accessing or exposing anything a `Plugin` developer needs in `WebSockets` with runtime communication between you and the server. And again this isn't `RPC-JSON` over `WebSockets`. This is like `SignalR` or `WebSockets` in general; type of communication. If we want `RPC-JSON` runtime communication we can just use `HTTP/2` or `HTTP/1.1` to do that. So why use `WebSockets` for something that can already be done with `HTTP/2` streaming` or `HTTP/1.1` `KeepAlive`?

Reply:
> @vncoelho 
> We are merging this repo to `neo` monorepo. Me and @Jim8y where going to do one library for all common logic. But after the `RPC` proposal changes _PR_. That's why I programed the logic kind of the same with `RpcServer` `RpcMethodName` attribute. So you can just add `WebSocketMethod` and be done. Also we can update all the other `Plugins` too; just by adding `WebSocketMethod` attribute.

Lets merge first. That is the best choice.
I am stuck with Consensus UT because of that.

Reply:
Anybody that do the PR merging the modules will be good, we just need to merge that asap to move forward on other topics as well.

Reply:
> We need a good solid reason why with pros and cons

Sure, it's as easy as having the same API over different transport. If we have JSON-RPC with one API, WS with another, then whatever else with third one then we'll have to maintain a set of incompatible things and people will have to think about these differences when creating their dApps.

> it's simpler and cheaper with bandwidth

No. Bandwidth costs a lot. Filtering events is easy, sending them over some link is much more overhead both for server and client. From what I've seen what most of dApp developers want is:
 * get notifications from their contracts
 * monitor transactions they send (acceptance/execution result)

Always pushing a stream of data that's 99% irrelevant for these use cases is a resource waste. Providing just an event stream over WS and resorting to a different JSON-RPC client for other requests is a headache.

> Some of those clients or users you talk about using RPC-JSON had to wait months for a bug fix to merge for https://github.com/neo-project/neo-modules/pull/812; that should of had a hotfix within two weeks or reasonable timeframe.

I agree with that, these problems should be solved much faster, but I'm not sure how adding yet another plugin solves the problem. To me it only makes things worse, it's just more code to maintain.

> It meant for accessing or exposing anything a Plugin developer

Well, it's plugin already, so to me, it should provide some value to the end user. If I install this plugin and the only thing I get is just an event stream then how do I integrate my dApp with it? Why should it be a part of the core? If you only need an unfiltered event stream for neo-express then it might as well be a plugin maintained as a part of neo-express. Just as an example, there is a https://github.com/Hecate2/neo-fairy-test/ maintained outside of the core.

Reply:
You do make good points. I understand completely. At 1st this PR was just `WebSocketServer` hints the name. Only acted as a `EventServer` for `node` events. See https://github.com/neo-project/neo/issues/2996 for full story.

>No. Bandwidth costs a lot. 

Here in the USA. It's almost free
![image](https://github.com/neo-project/neo-modules/assets/8141309/2b82b8e6-3a04-4d29-bd94-8feb03bc8687)

Reply:
Here in Brazil it is free. :D

Reply:
That's not the cost I'm talking about, who cares about those dollar bills... I'm more about the tx/rx overhead, latencies and channel capacity. Not transmitting even 1M of data is much much better than transmitting it, filtering events is some KBs of memory and nanoseconds of CPU time, communicating them to the client is milliseconds easily. Then there is a slow reader (client) problem which is much more likely to happen if event stream isn't filtered.

This approach is OK for some development environment where you have a node locally that's not doing a lot and just want to get everything from it. But this doesn't make for a good general-purpose plugin. So my suggestion is to either make it usable for dApps or move it out of the core.